contracts/MechManager.sol

```diff
@@ -0,0 +1,279 @@
+
+        // Traverse all the mech marketplaces and statuses
+        for (uint256 i = 0; i < mechMarketplaces.length; ++i) {
+            if (mechMarketplaces[i] == address(0)) {
+                revert ZeroAddress();
+            }
+
+            mapMechMarketplaces[mechMarketplaces[i]] = statuses[i];
+        }
+
+        emit SetMechMarketplaceStatuses(mechMarketplaces, statuses);
+    }
+
+    /// @dev Sets mech factory statues.
+    /// @param mechFactories Mech marketplace contract addresses.
+    /// @param statuses Corresponding whitelisting statues.
+    function setMechFactoryStatuses(address[] memory mechFactories, bool[] memory statuses) external {
+        // Check for the ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        if (mechFactories.length != statuses.length) {
+            revert WrongArrayLength(mechFactories.length, statuses.length);
+        }
+
+        // Traverse all the mech marketplaces and statuses
+        for (uint256 i = 0; i < mechFactories.length; ++i) {
+            if (mechFactories[i] == address(0)) {
+                revert ZeroAddress();
+            }
+
+            mapMechFactories[mechFactories[i]] = statuses[i];
+        }
+
+        emit SetMechFactoryStatuses(mechFactories, statuses);
+    }
+
+    function increaseRequestsCounts(address requester) external {
+        // Check for marketplace access
+        if (!mapMechMarketplaces[msg.sender]) {
+            revert UnauthorizedAccount(msg.sender);
+        }
+
+        // Record the request count
+        mapRequestCounts[requester]++;
+        // Increase the number of undelivered requests
+        numUndeliveredRequests++;
+        // Increase the total number of requests
+        numTotalRequests++;
+    }
+
+    function increaseDeliveryCounts(address requester, address agentMech, address mechServiceMultisig) external {
+        // Check for marketplace access
+        if (!mapMechMarketplaces[msg.sender]) {
+            revert UnauthorizedAccount(msg.sender);
+        }
+
+        // Decrease the number of undelivered requests
+        numUndeliveredRequests--;
+        // Increase the amount of requester delivered requests
+        mapDeliveryCounts[requester]++;
+        // Increase the amount of agent mech delivery counts
+        mapAgentMechDeliveryCounts[agentMech]++;
+        // Increase the amount of mech service multisig delivered requests
+        mapMechServiceDeliveryCounts[mechServiceMultisig]++;
+    }
+
+    /// @dev Validates agent mech.
+    /// @param agentMech Agent mech address.
+    /// @return status True, if the mech is valid.
+    function checkMechValidity(address agentMech) external view returns (bool status) {
+        // TODO: shall we also check the status of the factory?
+        // TODO: if yes, what if the factory was de-whitelisted? all the mechs then become invalid
+        status = mapAgentMechFactories[agentMech] != address(0);
+    }
```

Comment:
We do the same as in stkaing, once created its valid. Only check at creation time.

Reply:
Then anybody is able to create a factory with its mech and then deliver the request, so at least we ned to check created mechs.